### PR TITLE
docs: added tutorial on how to configure mTLS between the client and APISIX

### PR DESCRIPTION
**Pull request by guoqqqi**
### Description

Currently, our mTLS documentation is not user-friendly and users cannot quickly configure mTLS based on the documentation. We need to optimize it to make it clearer.
https://github.com/apache/apisix/blob/master/docs/en/latest/mtls.md

### What I do

Added tutorial on how to configure mTLS between the client and APISIX
- [x] zh

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: docs/zh/latest/tutorials/route-mtls.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
+---
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/route-mtls.md:1]: Slug: `/client-to-apisix-mtls`

File: docs/zh/latest/tutorials/route-mtls.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+TLS 双向认证一般简称为 mTLS，是目前主流的 TLS 协议。如果在你的网络环境中，要求只有受信任的客户端才可以访问服务端，那么 mTLS 是目前最合适的一种方案。本文主要介绍了如何配置 Apache APISIX 和客户端之间的 mTLS。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/route-mtls.md:29]: Check wiki

Comment (guoqqqi) [reply docs/zh/latest/tutorials/route-mtls.md]: update

File: docs/zh/latest/tutorials/route-mtls.md (hunk 1 above)
Comment (guoqqqi) [reply docs/zh/latest/tutorials/route-mtls.md]: done

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+openssl pkcs12 -export -clcerts -in client.cer -inkey client.key -out client.p12
+```
+
+### 在 APISIX 中配置证书
+
+使用 `curl` 命令请求 APISIX Admin API 创建一个 SSL 资源并指定 SNI。
+
+:::note 注意
+
+证书中的换行符需要替换为 `\n`
```
Comment (bzp2010) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:74]:
```suggestion
证书中的换行需要替换为其转义字符 `\n`
```

File: docs/zh/latest/mtls.md [hunk 4]
```diff
@@ -97,6 +97,8 @@ apisix:
 
 ### 如何配置
 
+我们提供了一个演示教程，详细的讲解了如何配置客户端和 APISIX 之间的 mTLS。[点击阅读](./tutorials//client-to-apisix-mtls.md)。
```
Comment (bzp2010) [inline docs/zh/latest/mtls.md:100]:
```suggestion
我们提供了一个演示教程，详细的讲解了如何配置客户端和 APISIX 之间的 mTLS。[点击阅读](./tutorials/client-to-apisix-mtls.md)。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+```shell
+# 127.0.0.1 localhost
+127.0.0.1 test.com
+```
+
+2. 重启网络模块
+
+```
+sudo /etc/init.d/networking restart
+```
```
Comment (bzp2010) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:146]: AFAIK, this is not necessary, hosts are not affected by the network manager in the OS distribution, it takes effect in real time.

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+^C
+--- test.com ping statistics ---
+4 packets transmitted, 4 received, 0% packet loss, time 3080ms
+rtt min/avg/max/mdev = 0.028/0.033/0.037/0.003 ms
+```
+
+4. 测试
+
+```shell
+curl https://test.com:9443/anything -k --cert ./client.cer --key ./client.key
```
Comment (bzp2010) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:167]: If we don't use hosts, we can modify the curl command to control its domain resolution behavior using the `--resolve` flag.

```sh
curl --resolve "test.com:9443:127.0.0.1" https://test.com:9443/anything -k --cert ./client.cer --key ./client.key

```

File: docs/zh/latest/mtls.md [hunk 7]
```diff
@@ -97,6 +97,8 @@ apisix:
 
 ### 如何配置
 
+我们提供了一个演示教程，详细的讲解了如何配置客户端和 APISIX 之间的 mTLS。[点击阅读](./tutorials/client-to-apisix-mtls.md)。
```
Comment (SylviaBABY) [inline docs/zh/latest/mtls.md:100]:
```suggestion
我们提供了一个[演示教程](./tutorials/client-to-apisix-mtls.md)，详细地讲解了如何配置客户端和 APISIX 之间的 mTLS。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+下面我们演示一个配置它的完整示例。
+
+主要包含以下过程：
+
+1. 生成证书；
+2. 在 APISIX 中配置证书；
+3. 在 APISIX 中创建并配置路由；
+4. 测试验证。
+
+为了使测试结果更加清晰，下面的演示中会向上游传递一些有关客户端证书的信息，其中包括：`serial`，`fingerprint` 和 `common name`。
```
Comment (SylviaBABY) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:42]:
```suggestion
为了使测试结果更加清晰，本文提到的示例会向上游传递一些有关客户端证书的信息，其中包括：`serial`，`fingerprint` 和 `common name`。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+### 生成证书
+
+我们需要生成三个测试证书，分别是根证书、服务器证书、客户端证书。只需要通过下面命令，它就可以通过 `OpenSSL` 生成我们需要的测试证书。
```
Comment (SylviaBABY) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:46]:
```suggestion
我们需要生成三个测试证书，分别是根证书、服务器证书、客户端证书。只需通过以下命令，就可以通过 `OpenSSL` 生成我们需要的测试证书。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            "httpbin.org":1
+        },
+        "type":"roundrobin"
+    }
+}'
+```
+
+APISIX 会根据 SNI 和上一步创建的 SSL 资源自动处理 TLS 握手，所以我们不需要在路由中指定主机名（如果你需要，你可以设置它）。
+
+另外，上面 `curl` 命令中，我们启用了 `proxy-rewrite` 插件，它将动态的更新请求头的信息，示例中变量值的来源是 `NGINX` 变量，你可以在这里找到它们：http://nginx.org/en/docs/http/ngx_http_ssl_module.html#variables。
```
Comment (SylviaBABY) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:127]:
```suggestion
另外，上面 `curl` 命令中，我们启用了 `proxy-rewrite` 插件，它将动态地更新请求头的信息，示例中变量值的来源是 `NGINX` 变量，你可以在这里找到它们：http://nginx.org/en/docs/http/ngx_http_ssl_module.html#variables。
```

**Review by SylviaBABY — DISMISSED** [review]

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
+---
+title: 配置路由的 mTLS
+keywords:
+  - mTLS
+  - API 网关
+  - APISIX
+description: 本文介绍了如何配置客户端和 Apache APISIX 之间的 mTLS。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:7]:
```suggestion
description: 本文介绍了如何在客户端和 Apache APISIX 之间配置双向认证（mTLS）。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
+---
+title: 配置路由的 mTLS
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:2]:
```suggestion
title: 配置客户端与 APISIX 之间的双向认证（mTLS）
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+TLS 双向认证一般简称为 mTLS，是一种身份验证的类型。如果在你的网络环境中，要求只有受信任的客户端才可以访问服务端，那么可以启用 mTLS 来验证消费者，保证服务端 API 的安全。本文主要介绍了如何配置 Apache APISIX 和客户端之间的 mTLS。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:29]:
```suggestion
mTLS 是一种双向身份认证的方式。如果在你的网络环境中，要求只有受信任的客户端才可以访问服务端，那么可以启用 mTLS 来验证客户端的身份，保证服务端 API 的安全。本文主要介绍了如何配置客户端与 Apache APISIX 之间的双向认证（mTLS）。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+## 配置
+
+下面我们演示一个配置它的完整示例。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:33]:
```suggestion
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+
+主要包含以下过程：
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:35]:
```suggestion
本示例包含包含以下过程：
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    "sni": "test.com",
+    "cert": "<服务器证书公钥>",
+    "key": "<服务器证书私钥>",
+    "client": {
+        "ca": "<客户端证书公钥>"
+    }
+}'
+```
+
+- `sni`：指定证书的域名（CN），当客户端尝试通过 TLS 与 APISIX 握手时，APISIX 会将 ClientHello 中的 SNI 数据与该字段进行匹配，找到对应的服务器证书进行握手。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:92]:
```suggestion
- `sni`：指定证书的域名（CN），当客户端尝试通过 TLS 与 APISIX 握手时，APISIX 会将 `ClientHello` 中的 SNI 数据与该字段进行匹配，找到对应的服务器证书进行握手。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+    "upstream": {
+        "nodes": {
+            "httpbin.org":1
+        },
+        "type":"roundrobin"
+    }
+}'
+```
+
+APISIX 会根据 SNI 和上一步创建的 SSL 资源自动处理 TLS 握手，所以我们不需要在路由中指定主机名（如果你需要，你可以设置它）。
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:125]:
```suggestion
APISIX 会根据 SNI 和上一步创建的 SSL 资源自动处理 TLS 握手，所以我们不需要在路由中指定主机名（但也可以显式地指定主机名）。
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+```shell
+# 根证书
+openssl genrsa -out ca.key 2048
+openssl req -new -sha256 -key ca.key -out ca.csr -subj "/CN=ROOTCA"
+openssl x509 -req -days 36500 -sha256 -extensions v3_ca -signkey ca.key -in ca.csr -out ca.cer
+
+# 服务器证书
+openssl genrsa -out server.key 2048
+openssl req -new -sha256 -key server.key -out server.csr -subj "/CN=test.com"
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:56]:
```suggestion
# 注意：CN 值中的 `test.com` 为我们要测试的域名/主机名。
openssl req -new -sha256 -key server.key -out server.csr -subj "/CN=test.com"
```

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+- 修改 /etc/hosts 文件
+
+  ```shell
+  # 127.0.0.1 localhost
+  127.0.0.1 test.com
+  ```
+
+- 验证测试域名是否生效
+
+  ```
```
Comment (juzhiyuan) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:150]:
```suggestion
  ```shell
```

**Review by juzhiyuan — APPROVED** [review]

**Review by SylviaBABY — APPROVED** [review]

**Comment by xinqing123** [comment]
步骤是一步一步来的，但是就是不行
image

**Comment by guoqqqi** [comment]
Hi, @xinqing123 Can you show me your router and ssl resource configuration information?

**Comment by xinqing123** [comment]
> 你好，@xinqing123你能告诉我你的路由器和ssl资源配置信息吗？

image
image

**Comment by xinqing123** [comment]
> 还请告诉我`uri`您的路线配置。
> 
> 顺便问一下，你用哪种方式运行 APISIX？
image

运行方式我就在linxu rpm安装，然后就启动了

**Comment by xinqing123** [comment]
> 

image

**Comment by xinqing123** [comment]
> It seems that the NSS version is too old to access https domains using `curl`.
> 
> use `yum -y update nss` to update `NSS`

image

**Comment by xinqing123** [comment]
> > It seems that the NSS version is too old to access https domains using `curl`.
> > use `yum -y update nss` to update `NSS`
> 
> image

我重启了apisix，然后报另一个错误
image

**Comment by xinqing123** [comment]
> > > 似乎 NSS 版本太旧，无法使用`curl`. 
> > > 用于`yum -y update nss`更新`NSS`
> > 
> > 
> > 图片
重启了apisix
image

**Comment by xinqing123** [comment]
成功了，非常感谢，爱你哦

**Pull request merged**